### ISSUE TITLE
Tailscale split DNS resolves addresses inconsistently

### ISSUE THREAD

**Issue by jinnko** [tailscale/tailscale#8294]
### What is the issue?

Tailscale DNS isn't honouring the Split DNS configuration 100% of the time.  Haven't identified a pattern to when it does and doesn't - seems to be random.

### Steps to reproduce

1. Configure Split DNS to resolve `something.mongodb.net` and `compute-1.amazonaws.com` via our AWS VPC DNS resolver.  The latter is required because the former returns a CNAME.

2. Query the Tailscale DNS server at `100.100.100.100` for `dbname-something.mongodb.net` and we get a public IP, e.g.:

       $ dig +short @100.100.100.100 ANY dbname-something.mongodb.net
       ec2-1-2-3-4.compute-1.amazonaws.com.
       1.2.3.4

3. Query the VPC DNS directly and we get an RFC1918 IP for the host:

       $ dig +short @100.100.100.100 ANY ec2-1-2-3-4.compute-1.amazonaws.com
       192.168.4.5

On some occasions step 2 returns the RFC1918 address, and on others it returns the public IP.  The VPC nameservers configured via the Tailscale DNS configuration always returns the RFC1918 IP.

### Are there any recent changes that introduced the issue?

Configured this last week and haven't seen it consistently work.  All clients are on at least tailscale client 1.40 and greater.

### OS

Linux, macOS

### OS version

Ubuntu 23.04, ArchLinux, macOS 13 Ventura

### Tailscale version

1.42.0 - across all OSs

### Other software

Subnet router is on the latest AmazonLinux also using 1.42.0 within the VPC.

### Bug report

BUG-d57f307c05581e74b4777ad65b11fba8b2b3ae87cea9c2f59ce982f9d14eec65-20230607124948Z-adeccef8ecd2218e

**Comment by licorsec** [comment]
I just wanted to confirm that we are experiencing the same issue on our Windows clients with version 1.44.0. As stated by jinnko, this seems to affect users at completely random intervals. At this point, I have not received any complaints from users on older versions. Our current workaround is having the user cycle the service using Tailscale down, tailscale up.

**Comment by floriantz** [comment]
We seem to possibly encounter a similar issue as well with up to date clients.

**Comment by floriantz** [comment]
@kelivel I had to revert the networking changes that enabled the behaviour sadly and it only occured on colleagues machines running Ubuntu 22+ which I don't so I can't reproduce it again for now 😞

**Comment by floriantz** [comment]
@kelivel Managed to reproduce the issue again by accident, hopefully that helps: 

BUG-859a7d26dc7355ed5b397b2c36fec3504714d2f216ddde662e6f91b8dcee950c-20231017135256Z-fa94d1717c7650ff

I was doing some dns resolution on the CNAME records that should end up giving me a RDS private IP, but sometime I randomly get the public IP instead.

**Comment by kelivel** [comment]
@floriantz I'd like to ask a few questions about your tailnet and I don't want to reveal your personal information, so would it be alright if I reached out to you directly using the email associated with your tailnet? Otherwise, reach me at http://www.tailscale.com/support and mention my username & a link to this issue, and we can look a bit closer.

**Comment by floriantz** [comment]
Yes it's alright to reach me at the owner email associated to my tailnet !

**Comment by phazeight** [comment]
experiencing something very similar using aws ec2 instance subnet router forwarding with SPLIT dns method to connect to RDS db via IAM SSO authentication and temp db access token. Sometimes works, sometimes doesn't. Working connection can be logged out of, then tried again minutes or seconds later, and will get connection timed out error.
